src/frontend/Pretty_printing.ml

```diff
@@ -369,63 +411,66 @@ and pp_recursive_ifthenelse ppf (s, loc) =
 
 and pp_statement ppf ({stmt= s_content; smeta= {loc}} as ss : untyped_statement)
     =
-  match s_content with
-  | Assignment {assign_lhs= l; assign_op= assop; assign_rhs= e} ->
-      pf ppf "@[<h>%a %a %a;@]" pp_lvalue l pp_assignmentoperator assop
-        pp_expression e
-  | NRFunApp (_, id, es) ->
-      pf ppf "%a(@[%a);@]" pp_identifier id pp_list_of_expression (es, loc)
-  | TargetPE e -> pf ppf "target += %a;" pp_expression e
-  | IncrementLogProb e ->
-      pf ppf "increment_log_prob(@[<hov>%a@]);" pp_expression e
-  | Tilde {arg= e; distribution= id; args= es; truncation= t} ->
-      pf ppf "%a ~ %a(@[%a)@]%a;" pp_expression e pp_identifier id
-        pp_list_of_expression (es, loc) pp_truncation t
-  | Break -> pf ppf "break;"
-  | Continue -> pf ppf "continue;"
-  | Return e -> pf ppf "return %a;" (hbox pp_expression) e
-  | ReturnVoid -> pf ppf "return;"
-  | Print ps -> pf ppf "print(%a);" pp_list_of_printables ps
-  | Reject ps -> pf ppf "reject(%a);" pp_list_of_printables ps
-  | Skip -> pf ppf ";"
-  | IfThenElse (_, _, _) ->
-      (vbox pp_recursive_ifthenelse) ppf (ss, ss.smeta.loc.begin_loc)
-  | While (e, s) -> pf ppf "while (%a) %a" pp_expression e pp_statement s
-  | For {loop_variable= id; lower_bound= e1; upper_bound= e2; loop_body= s} ->
-      pf ppf "@[<v>for (%a in %a : %a) %a@]" pp_identifier id pp_expression e1
-        pp_expression e2 pp_indent_unless_block (s, e2.emeta.loc.end_loc)
-  | ForEach (id, e, s) ->
-      pf ppf "for (%a in %a) %a" pp_identifier id pp_expression e
-        pp_indent_unless_block (s, e.emeta.loc.end_loc)
-  | Block vdsl ->
-      pf ppf "{@,%a@,}" (indented_box pp_list_of_statements) (vdsl, loc)
-  | Profile (name, vdsl) ->
-      pf ppf "profile(%s) {@,%a@,}" name
-        (indented_box pp_list_of_statements)
-        (vdsl, loc)
-  | VarDecl
-      { decl_type= pst
-      ; transformation= trans
-      ; identifier= id
-      ; initial_value= init
-      ; is_global= _ } ->
-      let pp_init ppf init =
-        match init with None -> () | Some e -> pf ppf " = %a" pp_expression e
-      in
-      let es =
-        match pst with
-        | Sized st -> Fn.compose snd unwind_sized_array_type st
-        | Unsized _ -> [] in
-      pf ppf "@[<h>%a%a %a%a;@]" pp_array_dims es pp_transformed_type
-        (pst, trans) pp_identifier id pp_init init
-  | FunDef {returntype= rt; funname= id; arguments= args; body= b} -> (
-      let loc_of (_, _, id) = id.id_loc in
-      pf ppf "%a %a(%a" pp_returntype rt pp_identifier id
-        (box (pp_list_of pp_args loc_of))
-        (args, {loc with end_loc= b.smeta.loc.begin_loc}) ;
-      match b with
-      | {stmt= Skip; _} -> pf ppf ");"
-      | b -> pf ppf ") %a" pp_statement b )
+  if (not !print_included) && Option.is_some loc.end_loc.included_from then ()
+  else
```

Comment:
```suggestion
  if !print_included || Option.is_none loc.end_loc.included_from then
```
The same line is also present in `pp_space`. Would it be possible to move all this logic to `pp_list_of_statements`? I'd rather have the caller decide which statements need to be printed than `pp_statement` and `pp_space` sometimes be no-ops. There are some calls to `pp_statement` outside of `pp_list_of_statements` but I don't think any of those allow #include anyway.

Reply:
I don't think we can move it because of the recursive calls inside `pp_statement`, like in the `while` case

Reply:
I don't think `while (...) #include` works either way.

Reply:
Ah yep, it only works if you put braces around it, which means it would be a Block and call the list version instead. 

I'm not able to move the logic out of pp_space still. It's still important to call pp_spac*ing* from inside list_of_statements in order to actually place the `#include`